File: mimir-backend/compatibility-tests/sdk-test-go/tests/neptune_test.go
package tests

import (
	"context"
	"fmt"
	"testing"
	"time"

	"mimir-sdk-test-go/internal/testutil"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/neptune"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNeptuneCluster(t *testing.T) {
	ctx := context.Background()
	svc := testutil.NeptuneClient()
	clusterId := fmt.Sprintf("go-neptune-%d", time.Now().UnixMilli()%100000)

	t.Cleanup(func() {
		svc.DeleteDBCluster(ctx, &neptune.DeleteDBClusterInput{ //nolint:errcheck
			DBClusterIdentifier: aws.String(clusterId),
			SkipFinalSnapshot:   aws.Bool(true),
		})
	})

	t.Run("CreateDBCluster", func(t *testing.T) {
		out, err := svc.CreateDBCluster(ctx, &neptune.CreateDBClusterInput{
			DBClusterIdentifier: aws.String(clusterId),
			Engine:              aws.String("neptune"),
		})
		require.NoError(t, err)
		require.NotNil(t, out.DBCluster)
		assert.Equal(t, clusterId, aws.ToString(out.DBCluster.DBClusterIdentifier))
		assert.Equal(t, "neptune", aws.ToString(out.DBCluster.Engine))
		assert.Equal(t, "available", aws.ToString(out.DBCluster.Status))
		assert.Contains(t, aws.ToString(out.DBCluster.DBClusterArn), "arn:aws:neptune:")
		assert.Greater(t, aws.ToInt32(out.DBCluster.Port), int32(0))
	})

	t.Run("DescribeDBClusters", func(t *testing.T) {
		out, err := svc.DescribeDBClusters(ctx, &neptune.DescribeDBClustersInput{
			DBClusterIdentifier: aws.String(clusterId),
		})
		require.NoError(t, err)
		require.Len(t, out.DBClusters, 1)
		assert.Equal(t, clusterId, aws.ToString(out.DBClusters[0].DBClusterIdentifier))
	})

	t.Run("ModifyDBCluster", func(t *testing.T) {
		out, err := svc.ModifyDBCluster(ctx, &neptune.ModifyDBClusterInput{
			DBClusterIdentifier:             aws.String(clusterId),
			EnableIAMDatabaseAuthentication: aws.Bool(true),
		})
		require.NoError(t, err)
		assert.True(t, aws.ToBool(out.DBCluster.IAMDatabaseAuthenticationEnabled))
	})

	t.Run("DeleteDBCluster", func(t *testing.T) {
		_, err := svc.DeleteDBCluster(ctx, &neptune.DeleteDBClusterInput{
			DBClusterIdentifier: aws.String(clusterId),
			SkipFinalSnapshot:   aws.Bool(true),
		})
		require.NoError(t, err)
	})
}

func TestNeptuneInstance(t *testing.T) {
	ctx := context.Background()
	svc := testutil.NeptuneClient()
	clusterId := fmt.Sprintf("go-neptune-cl-%d", time.Now().UnixMilli()%100000)
	instanceId := fmt.Sprintf("go-neptune-inst-%d", time.Now().UnixMilli()%100000)

	// Create cluster first
	_, err := svc.CreateDBCluster(ctx, &neptune.CreateDBClusterInput{
		DBClusterIdentifier: aws.String(clusterId),
		Engine:              aws.String("neptune"),
	})
	require.NoError(t, err, "setup: create Neptune cluster")

	t.Cleanup(func() {
		svc.DeleteDBInstance(ctx, &neptune.DeleteDBInstanceInput{ //nolint:errcheck
			DBInstanceIdentifier: aws.String(instanceId),
		})
		svc.DeleteDBCluster(ctx, &neptune.DeleteDBClusterInput{ //nolint:errcheck
			DBClusterIdentifier: aws.String(clusterId),
			SkipFinalSnapshot:   aws.Bool(true),
		})
	})

	t.Run("CreateDBInstance", func(t *testing.T) {
		out, err := svc.CreateDBInstance(ctx, &neptune.CreateDBInstanceInput{
			DBInstanceIdentifier: aws.String(instanceId),
			DBClusterIdentifier:  aws.String(clusterId),
			DBInstanceClass:      aws.String("db.r5.large"),
			Engine:               aws.String("neptune"),
		})
		require.NoError(t, err)
		require.NotNil(t, out.DBInstance)
		assert.Equal(t, instanceId, aws.ToString(out.DBInstance.DBInstanceIdentifier))
		assert.Equal(t, clusterId, aws.ToString(out.DBInstance.DBClusterIdentifier))
		assert.Equal(t, "available", aws.ToString(out.DBInstance.DBInstanceStatus))
	})

	t.Run("DescribeDBInstances", func(t *testing.T) {
		out, err := svc.DescribeDBInstances(ctx, &neptune.DescribeDBInstancesInput{
			DBInstanceIdentifier: aws.String(instanceId),
		})
		require.NoError(t, err)
		require.Len(t, out.DBInstances, 1)
		assert.Equal(t, instanceId, aws.ToString(out.DBInstances[0].DBInstanceIdentifier))
	})

	t.Run("ModifyDBInstance", func(t *testing.T) {
		out, err := svc.ModifyDBInstance(ctx, &neptune.ModifyDBInstanceInput{
			DBInstanceIdentifier: aws.String(instanceId),
			DBInstanceClass:      aws.String("db.r5.xlarge"),
		})
		require.NoError(t, err)
		assert.Equal(t, "db.r5.xlarge", aws.ToString(out.DBInstance.DBInstanceClass))
	})

	t.Run("DeleteDBInstance", func(t *testing.T) {
		_, err := svc.DeleteDBInstance(ctx, &neptune.DeleteDBInstanceInput{
			DBInstanceIdentifier: aws.String(instanceId),
		})
		require.NoError(t, err)
	})
}

File: mimir-backend/compatibility-tests/sdk-test-rust/tests/neptune_test.rs
mod common;

#[tokio::test(flavor = "multi_thread")]
async fn test_neptune_create_cluster() {
    let neptune = common::neptune_client().await;
    let cluster_id = "rust-neptune-create-cluster";

    let result = neptune
        .create_db_cluster()
        .db_cluster_identifier(cluster_id)
        .engine("neptune")
        .send()
        .await;
    assert!(result.is_ok(), "CreateDBCluster failed: {:?}", result.err());

    let cluster = result.unwrap().db_cluster.unwrap();
    assert_eq!(cluster.db_cluster_identifier().unwrap_or(""), cluster_id);
    assert_eq!(cluster.engine().unwrap_or(""), "neptune");
    assert_eq!(cluster.status().unwrap_or(""), "available");
    assert!(cluster.db_cluster_arn().unwrap_or("").starts_with("arn:aws:neptune:"));

    let _guard = common::CleanupGuard::new({
        let neptune = neptune.clone();
        async move {
            let _ = neptune
                .delete_db_cluster()
                .db_cluster_identifier(cluster_id)
                .skip_final_snapshot(true)
                .send()
                .await;
        }
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn test_neptune_describe_cluster() {
    let neptune = common::neptune_client().await;
    let cluster_id = "rust-neptune-describe-cluster";

    neptune
        .create_db_cluster()
        .db_cluster_identifier(cluster_id)
        .engine("neptune")
        .send()
        .await
        .expect("setup: create cluster");

    let _guard = common::CleanupGuard::new({
        let neptune = neptune.clone();
        async move {
            let _ = neptune
                .delete_db_cluster()
                .db_cluster_identifier(cluster_id)
                .skip_final_snapshot(true)
                .send()
                .await;
        }
    });

    let result = neptune
        .describe_db_clusters()
        .db_cluster_identifier(cluster_id)
        .send()
        .await;
    assert!(result.is_ok(), "DescribeDBClusters failed: {:?}", result.err());

    let clusters = result.unwrap().db_clusters;
    assert_eq!(clusters.len(), 1);
    assert_eq!(clusters[0].db_cluster_identifier().unwrap_or(""), cluster_id);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_neptune_modify_cluster() {
    let neptune = common::neptune_client().await;
    let cluster_id = "rust-neptune-modify-cluster";

    neptune
        .create_db_cluster()
        .db_cluster_identifier(cluster_id)
        .engine("neptune")
        .send()
        .await
        .expect("setup: create cluster");

    let _guard = common::CleanupGuard::new({
        let neptune = neptune.clone();
        async move {
            let _ = neptune
                .delete_db_cluster()
                .db_cluster_identifier(cluster_id)
                .skip_final_snapshot(true)
                .send()
                .await;
        }
    });

    let result = neptune
        .modify_db_cluster()
        .db_cluster_identifier(cluster_id)
        .enable_iam_database_authentication(true)
        .send()
        .await;
    assert!(result.is_ok(), "ModifyDBCluster failed: {:?}", result.err());

    let cluster = result.unwrap().db_cluster.unwrap();
    assert!(cluster.iam_database_authentication_enabled());
}

#[tokio::test(flavor = "multi_thread")]
async fn test_neptune_create_and_describe_instance() {
    let neptune = common::neptune_client().await;
    let cluster_id = "rust-neptune-instance-cluster";
    let instance_id = "rust-neptune-instance";

    neptune
        .create_db_cluster()
        .db_cluster_identifier(cluster_id)
        .engine("neptune")
        .send()
        .await
        .expect("setup: create cluster");

    let _cluster_guard = common::CleanupGuard::new({
        let neptune = neptune.clone();
        async move {
            let _ = neptune
                .delete_db_cluster()
                .db_cluster_identifier(cluster_id)
                .skip_final_snapshot(true)
                .send()
                .await;
        }
    });

    let create = neptune
        .create_db_instance()
        .db_instance_identifier(instance_id)
        .db_cluster_identifier(cluster_id)
        .db_instance_class("db.r5.large")
        .engine("neptune")
        .send()
        .await;
    assert!(create.is_ok(), "CreateDBInstance failed: {:?}", create.err());

    let instance = create.unwrap().db_instance.unwrap();
    assert_eq!(instance.db_instance_identifier().unwrap_or(""), instance_id);
    assert_eq!(instance.db_cluster_identifier().unwrap_or(""), cluster_id);
    assert_eq!(instance.db_instance_status().unwrap_or(""), "available");

    let _inst_guard = common::CleanupGuard::new({
        let neptune = neptune.clone();
        async move {
            let _ = neptune
                .delete_db_instance()
                .db_instance_identifier(instance_id)
                .send()
                .await;
        }
    });

    let describe = neptune
        .describe_db_instances()
        .db_instance_identifier(instance_id)
        .send()
        .await;
    assert!(describe.is_ok(), "DescribeDBInstances failed: {:?}", describe.err());
    assert_eq!(describe.unwrap().db_instances.len(), 1);
}

